### ISSUE TITLE
Allow plugins to specify file extensions that they support

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#8920]
I'm creating this issue to follow up on the TSC discussion [here](https://github.com/eslint/tsc-meetings/blob/df5bb3c39c6a87f0e11dbb1e5c0fda7f54563972/notes/2017/2017-05-11.md#breaking-add-jsx-to-the-default---ext-file-extensions-fixes-8399-8515) about allowing plugins to specify what file extensions they support. The problem at the moment is that it's tedious for people to use other file extensions (such as `.jsx`, `.md`, and `.html`) because they have to use a command-line flag to parse the files at all.

## Proposal

This would require two small changes to ESLint core:

1. Allow plugins to specify a `supportedFileExtensions` property containing an array of strings, defaulting to `['.js']`. Use that to determine whether a given file should be parsed. (Since only one parser can be active at a time, only the option from the highest-precedence plugin would get used, rather than merging the array from multiple plugins.)
1. Pass the file extension to parsers as part of the options parameter.

The idea of (2) is that people could create "wrapper parsers" like this:

```js
// parser that handles multiple file types
module.exports = {
    parse(text, options) {
        if (options.fileExtension === "js") {
            return javascriptParser(text, options);
        }
        if (options.fileExtension === "jsx") {
            return jsxParser(text, options);
        }
        // ...
    }
};
```

## Drawbacks

Currently, we filter a file using the `--ext` option before we resolve the config for that file. With this change, we would need to resolve configs for every file in a folder, even the ones that don't end up matching the supported extensions.

**Comment by mysticatea** [comment]
About 1, sounds good to me. Also, I think custom parsers should have the same option. For example, [typescript-eslint-parser](https://github.com/eslint/typescript-eslint-parser) supports `.ts` files and [vue-eslint-parser](https://github.com/mysticatea/vue-eslint-parser) supports `.vue` files. If a plugin has the option, it's prior to custom parser's since preprocessors runs before parsers.

About 2, I think it's unnecessary since [`options.filePath` includes extensions](https://github.com/mysticatea/vue-eslint-parser/blob/master/lib/parse-component.js#L143-L145).

**Comment by ilyavolodin** [comment]
Just wanted a clarification on point 1: 
> Allow *plugins* to specify a ...
> Since only one *parser* can be active at a time,
So are we talking about plugins or parsers? If plugins, then don't forget about preprocessors. We wanted to rewamp them for a while to be able to specify file formats that they support.

**Comment by not-an-aardvark** [comment]
Both of those are correct as written. The proposal updates plugins rather than parsers because it needs to work with preprocessors, but the value from the plugin would not get merged with other plugins because only one parser can be loaded at a  time.

**Comment by ilyavolodin** [comment]
I'm not sure that's a good idea. Since multiple preprocessors can work with the same parser, but different file extension (for example eslint-plugin-markdown and eslint-plugin-react).

**Comment by mysticatea** [comment]
I challenge to clear.

- Plugins can add supported extensions by preprocessors. Preprocessors are registered for each extension, so multiple preprocessors from multiple plugins can live in together at a time. `Object.keys(plugin.processors)` would be the supported extensions, and those should be merged from all plugins.
- Separately, custom parsers can add supported extensions by the `parse` function. It enables only one parser. I think that we need to support `parser.getSupportedExtensions(parserOptions)` for custom parsers.
- Then, it's useful if ESLint uses those supported extensions of plugins and custom parsers as `--ext`.

**Comment by not-an-aardvark** [comment]
After looking at this again, I agree that processors are a good way to handle this.

**Comment by not-an-aardvark** [comment]
Closing because this proposal will probably need to be changed significantly before we consider it again.
